src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java

```diff
@@ -93,6 +97,112 @@ public ExerciseService(ExerciseRepository exerciseRepository, ExerciseUnitReposi
+     * @param user      user
+     * @return subset of the exercises that a user allowed to see
+     */
+    public Set<Exercise> filterOutExercisesThatUserShouldNotSee(Set<Exercise> exercises, User user) {
+        if (exercises == null || user == null || exercises.isEmpty()) {
+            return Set.of();
+        }
+        List<Course> courses = exercises.stream().map(Exercise::getCourseViaExerciseGroupOrCourseMember).collect(Collectors.toList());
+        if (courses.size() != 1) {
+            throw new IllegalArgumentException("All exercises must be from the same course!");
+        }
+        Course course = courses.get(0);
+
+        Set<Exercise> exercisesUserIsAllowedToSee = new HashSet<>();
+        if (authCheckService.isAtLeastTeachingAssistantInCourse(course, user)) {
+            exercisesUserIsAllowedToSee = exercises;
+        }
+        else if (authCheckService.isStudentInCourse(course, user)) {
+            for (Exercise exercise : exercises) {
+                if (course.isOnlineCourse()) {
+                    // students in online courses can only see exercises where the lti outcome url exists, otherwise the result cannot be reported later on
+                    Optional<LtiOutcomeUrl> ltiOutcomeUrlOptional = ltiOutcomeUrlRepository.findByUserAndExercise(user, exercise);
+                    if (ltiOutcomeUrlOptional.isPresent()) {
+                        exercisesUserIsAllowedToSee.add(exercise);
+                    }
+                }
+                else {
+                    exercisesUserIsAllowedToSee.add(exercise);
+                }
+            }
+            exercisesUserIsAllowedToSee = exercisesUserIsAllowedToSee.stream().filter(Exercise::isVisibleToStudents).collect(Collectors.toSet());
+        }
+        return exercisesUserIsAllowedToSee;
+    }
+
+    /**
+     * Loads exercises with all the necessary information to display them correctly in the Artemis dashboard
+     *
+     * @param exerciseIds exercises to load
+     * @param user        user to load exercise information for
+     * @return exercises with all the necessary information loaded for correct display in the Artemis dashboard
+     */
+    public Set<Exercise> loadExercisesWithInformationForDashboard(Set<Long> exerciseIds, User user) {
+        if (exerciseIds == null || user == null) {
+            throw new IllegalArgumentException();
+        }
+        if (exerciseIds.isEmpty()) {
+            return new HashSet<>();
+        }
+        Set<Exercise> exercises = exerciseRepository.findByExerciseIdWithCategories(exerciseIds);
+        Set<Course> courses = exercises.stream().map(Exercise::getCourseViaExerciseGroupOrCourseMember).collect(Collectors.toSet());
+        if (courses.size() != 1) {
+            throw new IllegalArgumentException("All exercises must be from the same course!");
+        }
+        Course course = exercises.stream().findFirst().get().getCourseViaExerciseGroupOrCourseMember();
```

Comment:
This again can be `... = courses.get(0)`, can't it?

Reply:
I guess you mean exercises.get(0)

Reply:
We know all exercises belong to the same course so we can get the first course in the list we assembled to get the relevant course.

Reply:
Changed in 1731fed7a02c651ee44973f34f3d276b1aac5cd1